### ISSUE TITLE
Suggestion: Model Class

### ISSUE THREAD

**Issue by SolarLune** [GoranM/bdx#610]
Hey, there. 

So Goran suggested a Model class, which would be used to modify a GameObject's modelInstance easily and quickly. I'd like it to handle vertex manipulation and UV sliding easily and quickly. Not sure if this would require a Vertex class as well...

Any ideas for what it should do / how it should work?

**Comment by rafcolson** [comment]
Naturally `GameObject.model` would be an instance of a LibGdx Model extension. The `Model` class would contain methods for altering meshes, mesh parts, vertices, ... But what kind of alterations would we talk about then? Apart from uv scrolling, I can only think of armature deformation and texture deformation. Methods on GameObject for replacing the model could become ``Model.replace(model)`. Also the method to make the model unique becomes ```Model.unique()```.

It would make more sense to me if materials are accessed through the Model instead of the GameObject. Maybe we could extend `Array<...g3d.Material>` to `ArrayMaterials`? I'm not sure.

**Comment by GoranM** [comment]
> So Goran suggested a Model class, which would be used to modify a GameObject's modelInstance

I think the vertex buffer is managed by Model, not ModelInstance, so we would need a class that extends Model, not ModelInstance.

@rafcolson 

Your suggestions seem reasonable, but I don't think they should be a part of the initial patch. For now, it would be best to just have a basic extension of Model, with simple methods to manipulate vertex data.

**Comment by SolarLune** [comment]
If we just extend Model, then the modelInstance field on GameObject won't work, right? Don't we need to extend at least ModelInstance for it to be created correctly / for the added mesh altering functions to be accessible?

**Comment by GoranM** [comment]
I don't really understand what you mean.

The data is on the Model. You would get a reference to a model, and the methods would be accessible from there.

**Comment by SolarLune** [comment]
So we wouldn't have a reference to a ModelInstance anymore, it'd just be a reference to our Model. OK, I think I see. 

So what's the list of things that our Model class would need?

Like, would we have `moveVertex()` and `uvVertex()` functions, or something like that? Or just a list of Vertex instances representing vertices on the model that the user can move or adjust...?

To sum up, I made this initially for UV sliding across the entire mesh, so something that does that, in addition to UV adjustment for each individual vertex, would probably be cool.

**Comment by GoranM** [comment]
> So we wouldn't have a reference to a ModelInstance anymore, it'd just be a reference to our Model. OK, I think I see.

No, we would still need to have a ModelInstance. In blender terms you can think of ModelInstance as the Object, and Model as the Mesh.

As for what the API could look like:

```java
model.setVertPos(idx, x, y, z);
model.setVertUV(idx, u, v);
model.translateUVs(du, dv);
```

Something like that would probably be a good start.

**Comment by SolarLune** [comment]
I still don't quite understand, then, how exactly we would have access to the Model. Through the ModelInstance?

**Comment by GoranM** [comment]
Yes.

**Comment by SolarLune** [comment]
But if we use LibGDX's ModelInstance class, the user will have to cast its model field down to our BDXModel class whenever they refer to it.

**Comment by GoranM** [comment]
I don't think that would be too troublesome, but I guess we could also have a Model field on the GameObject.

Either would be fine, in my view.

**Comment by SolarLune** [comment]
Hmm, OK, so:

Basically, there doesn't appear to be any easy way to extend a custom Model class from the existing LibGDX one. ModelBuilder always returns a finished Model, and it doesn't appear to be possible to grab model data in a usable way (not really, anyway). In the createModel function, when I tried to pass the nodes, meshes, etc. from the finished model to the BDXModel and return that, BDX crashes. A fatal error having been detected by the JRE - an exception access violation.

So before I continue, I wonder if an easier way would be to continue the kind of paradigm that we have going for some of our classes and just have BDXModel "wrap" around the LibGDX ModelInstance. We could have each GameObject keep a reference to the BDXModel instead of the ModelInstance this way, allow you to modify the model through its functions rather than the ModelInstance's Model's MeshPart's Mesh's functions (or whatever it is). 

Basically, we'd push the ModelInstance into a BDXModel class, and put our functions there.

**Comment by GoranM** [comment]
Wrapping would work, but I don't see why we should wrap ModelInstance, because data lives on the Model.

**Comment by SolarLune** [comment]
Well, we could wrap the Model, but I thought it'd make more sense to wrap the ModelInstance because the ModelInstance references the Model, and the ModelInstance is what the mesh creation and rendering processes use. If we wrap the Model, then the engine'll be dealing with two different things (the ModelInstance and our BDXModel class). 

We'd also have to add some sort of `getModel()` function that returns the model, whereas if we wrapped the ModelInstance, then we could just make the new BDXModel field public and let users access the functions directly. It's up to you, though.

**Comment by GoranM** [comment]
> I thought it'd make more sense to wrap the ModelInstance because the ModelInstance references the Model, and the ModelInstance is what the mesh creation and rendering processes use.

As far as I know, our "mesh creation" process is about converting blender meshes into libgdx models. The renderer takes ModelInstances, but aside from the transform, most of the relevant data used in the render process is on the model.

> If we wrap the Model, then the engine'll be dealing with two different things (the ModelInstance and our BDXModel class).

Well, logically, they are "two different things", so that would make sense; Just as there's a split between Object and Mesh in blender, there is a corresponding split between ModelInstance and Model.

> if we wrapped the ModelInstance, then we could just make the new BDXModel field public and let users access the functions directly.

But doesn't "BDXModel" imply that you're wrapping libgdx Model?

Wrapping both is fine. As long as the methods for modifying vertex data are on the Model, I think that would be logically consistent.

**Comment by SolarLune** [comment]
> 
Well, logically, they are "two different things", so that would make sense; Just as there's a split between Object and Mesh in blender, there is a corresponding split between ModelInstance and Model.

**Comment by SolarLune** [comment]
> As far as I know, our "mesh creation" process is about converting blender meshes into libgdx models. The renderer takes ModelInstances, but aside from the transform, most of the relevant data used in the render process is on the model.

My mistake, I meant that the mesh creation process in BDX returns a Model, which we make ModelInstances of for each GameObject. Each GameObject doesn't know anything about a Model, but rather the ModelInstance (which contains a reference to the Model).

> Well, logically, they are "two different things", so that would make sense; Just as there's a split between Object and Mesh in blender, there is a corresponding split between ModelInstance and Model.

Well, not quite; it is similar in that a ModelInstance is an instance of a Model at a position, like an Object uses an instance of a Mesh at a position. However, a ModelInstance isn't an Object. In BDX, a GameObject has a ModelInstance. For all intents and purposes, the ModelInstance is, essentially, the mesh.

> But doesn't "BDXModel" imply that you're wrapping libgdx Model?

We could name it "BDXModelInstance", or whatever the case. "BDXModel" as a name just says that it's a model, not that it's our equivalent of the LibGDX Model (like how our Light isn't a direct coorelation to LibGDX's Lights, as ours can be any type).

> Wrapping both is fine. As long as the methods for modifying vertex data are on the Model, I think that would be logically consistent.

I don't really see the need to wrap both as wrapping is just to have the data in one place (it's not like we're extending the Model or ModelInstance). The functions we're going to add don't have to be in any particular location.

**Comment by rafcolson** [comment]
>For all intents and purposes, the ModelInstance is, essentially, the mesh.

I don't think so. The ModelInstance is the object really. Our GameObject manages the ModelInstance and the Body. The Model is much like the mesh in Blender.

**Comment by GoranM** [comment]
> However, a ModelInstance isn't an Object.

It corresponds directly to our notion of Object, just as Model corresponds directly to our notion of Mesh.

> In BDX, a GameObject has a ModelInstance. For all intents and purposes, the ModelInstance is, essentially, the mesh.

I don't really understand how the fact that every GameObject has a ModelInstance implies that a ModelInstance is essentially the mesh.

Assuming that you can provide some support for that statement: If a ModelInstance is essentially the Mesh, what does Model represent?

**Comment by SolarLune** [comment]
> Assuming that you can provide some support for that statement: If a ModelInstance is essentially the Mesh, what does Model represent?

The ModelInstance, coupled with the Model, represents a Mesh in Blender. You wouldn't really use one without the other, and while I know the Object in Blender corresponds to a ModelInstance in that it's the location that a Mesh is used, Objects in Blender are directly turned into GameObjects in BDX.

In any case, there's no point in discussing this further; this is just semantics more or less at this point. You guys seem to agree on this, so I'll go along with it. 

So the general concensus is to give each GameObject a "BDXModel" that references the Model, and allow mesh alteration operations through that instance, right?

**Comment by GoranM** [comment]
It's important to understand that BDX GameObject *wraps* LibGDX ModelInstance, it doesn't combine ModelInstance and Model into some composite notion that corresponds to a blender mesh - The Model corresponds to a blender Mesh.

I don't agree that it's "just semantics".

> So the general concensus is to give each GameObject a "BDXModel" that references the Model, and allow mesh alteration operations through that instance, right?

I think that would make sense. I assume @rafcolson holds the same view.

Just as GameObject wraps ModelInstance, we could have a Mesh class that wraps Model.

**Comment by rafcolson** [comment]
>I assume @rafcolson holds the same view.

As the API states `GameObject.modelInstance` was just a place to go if you needed to alter the `GameObject`'s mesh in a way that BDX does not provide. With a `GameObject.mesh` field, `GameObject.modelInstance` will stop being that place. (Excepts for Animations, until Armatures are up and running). So yes, I like the idea of a `Mesh` class.

**Comment by SolarLune** [comment]
Hmm. OK, I think I see.

[Here's](https://github.com/SolarLune/bdx/tree/ModelClass) the Mesh class. I think everything's working, but I'm not so sure about the matrix transpose functions (I rarely use Matrices, so I'm not sure if they're correct). Please feel free to test it all out and look over the API.

**Comment by GoranM** [comment]
I don't see any transpose functions ... Maybe you meant transform functions?

If you have working code, you should make a PR, and we should continue the discussion there.

**Comment by SolarLune** [comment]
Yeah, sorry, the transform functions.

OK, PR created.

**Comment by SolarLune** [comment]
OK! This was merged, so I'll close this. Thanks, guys.
